crates/ra_parser/src/grammar/items.rs

```diff
@@ -121,7 +121,13 @@ pub(super) fn maybe_item(p: &mut Parser, m: Marker, flavor: ItemFlavor) -> Resul
             T![unsafe] => {
                 // test default_unsafe_impl
                 // default unsafe impl Foo {}
-                if p.nth(2) == T![impl] {
+
+                // test default_unsafe_fn
+                // impl T for Foo {
+                //     default unsafe fn foo() {}
+                // }
+                let sk = p.nth(2);
+                if sk == T![impl] || sk == T![fn] {
```

Comment:
```suggestion
                if p.nth(2) == T![impl] || p.nth(2) == T![fn] {
```

We generally don't extract calls to nth.

Reply:
Let's hope I edited the code correctly...

Reply:
> We generally don't extract calls to nth.

Do we have true lookahead here? I remember some quadratic behaviour around MBE expansion and parsing.

Reply:
I think this should be constant-time yes. At some point, we should actually profile the stuff and make sure that the parser's code is tight, but, even than, I hope we'll convince the optimizer to CSE lokaheads...

Reply:
I was thinking of #1648 and #1643.